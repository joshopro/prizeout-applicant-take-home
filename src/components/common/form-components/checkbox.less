@import '../../../utils/styles/mixins.less';

.checkbox {
	.flexbox(@flexFlow: row nowrap; @justifyContent: flex-start; @alignItems: flex-start);
	
	position: relative;
	margin: .625rem 0;

	&:focus-within {
		.checkbox__control:after {
			border: 2px solid var(--color-primary);
		}
	}

	&__input {
		/* Hide real input and use fake controller to style */
		&[type="checkbox"]:not(:checked),
		&[type="checkbox"]:checked {
			position: absolute;
			opacity: 0;
			pointer-events: none;
		}

		/* Checked styles */
		&[type="checkbox"]:checked {
			+ .checkbox__control {
				/* Show the check when checked */
				&:before {
					opacity: 1;
				}
			}
		}

		/* Disabled styles */
		&[type="checkbox"]:not(:checked),
		&[type="checkbox"]:checked {
			&:disabled + .checkbox__control {
				opacity: .5;

				&:after {
					border: 1px solid rgba(2, 20, 49, 0.1);
					cursor: not-allowed;
				}
			}
		}

		/* Disabled styles when not checked */
		&[type="checkbox"]:not(:checked) {
			&:disabled + .checkbox__control {
				&:before {
					opacity: 0;
				}
			}
		}

		/* Disabled styles when checked */
		&[type="checkbox"]:checked {
			&:disabled + .checkbox__control {
				&:before {
					opacity: 1;
					color: var(--color-tertiary);
				}
			}
		}
	}
	
	/* Fake checkbox controller */
	&__control {
		position: relative;
		padding-left: 1.5rem;
		display: inline-block;
		user-select: none;

		/* Style the checkbox */
		&:before,
		&:after {
			content: '';
			position: absolute;
			top: var(--unit);
			left: 0;
			width: 1rem;
			height: 1rem;
			z-index: 0;
			border-radius: 1px;
			margin-top: 0;
			cursor: pointer;
		}

		/* Style the check, hide when not checked */
		&:before {
			.transition(@duration: .2s; @prop: opacity;);

			opacity: 0;
			content: '\f00c';
			font-family: var(--font-family-fa);
			color: var(--color-primary);
			font-size: 0.625rem;
			border: none;
			left: 3px;
			top: var(--unit);
			
			@media (prefers-reduced-motion: reduce) {
				transition: none;
			}
		}

		/* Default checkbox border color */
		&:after {
			border: 2px solid rgba(2, 20, 49, 0.25);
			border-radius: var(--unit);
		}
	}
}
