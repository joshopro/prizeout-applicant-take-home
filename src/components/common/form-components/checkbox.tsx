import React, { Fragment, useState } from 'react';
import PropTypes from 'prop-types';
import { v4 as uuid_v4 } from 'uuid';

import './checkbox.less';

interface CheckboxProps {
    content: string;
    id?: string;
    isRequired?: boolean;
    register?: Record<string, unknown>;
    title: string;
}

export const Checkbox: React.FC<CheckboxProps> = ({ content, id, isRequired, register, title }): React.ReactElement => {
    const [isChecked, setIsChecked] = useState(false),
        toggleChecked = (): void => setIsChecked(!isChecked),
        elmId = id || `checkbox-${uuid_v4()}`;

    return (
        <Fragment>
            <label data-testid="checkbox-label" className="checkbox" htmlFor={elmId}>
                <input
                    data-testid="checkbox-input"
                    id={elmId}
                    className="checkbox__input"
                    type="checkbox"
                    title={title}
                    onClick={toggleChecked}
                    {...register}
                    aria-required={isRequired}
                    defaultChecked={isChecked}
                />
                <span
                    data-testid="checkbox-input-control"
                    className="checkbox__control"
                    aria-checked={isChecked}
                    aria-labelledby={elmId}
                >
                    <small data-testid="checkbox-description" dangerouslySetInnerHTML={{ __html: content }}></small>
                </span>
            </label>
        </Fragment>
    );
};

Checkbox.propTypes = {
    content: PropTypes.string.isRequired,
    id: PropTypes.string,
    isRequired: PropTypes.bool,
    register: PropTypes.any,
    title: PropTypes.string,
};
