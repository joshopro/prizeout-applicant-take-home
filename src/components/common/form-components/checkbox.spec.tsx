import { render, screen } from '../../../testing/test-utils';
import userEvent from '@testing-library/user-event';
import React from 'react';
import '@testing-library/jest-dom';

import { Checkbox } from './checkbox';

describe('Test Checkbox Component', () => {
    const content = `Checkbox content`;
    const id = `checkbox-test-id`;
    const isRequired = true;
    const title = `Checkbox content`;

    // Matches snapshot
    test('Component matches snapshot', () => {
        const { asFragment } = render(<Checkbox content={content} id={id} isRequired={isRequired} title={title} />);

        expect(asFragment()).toMatchSnapshot();
    });

    // Checkbox has description
    test('Checkbox description is displayed', () => {
        render(<Checkbox content={content} id={id} isRequired={isRequired} title={title} />);

        expect(screen.getByText(content)).toBeVisible();
    });

    // Checkbox has type checkbox
    test('Checkbox input has type checkbox', () => {
        render(<Checkbox content={content} id={id} isRequired={isRequired} title={title} />);

        expect(screen.getByTestId('checkbox-input')).toHaveAttribute('type', expect.stringContaining('checkbox'));
    });

    // class of "checkbox"
    test('Checkbox has a class of checkbox', () => {
        render(<Checkbox content={content} id={id} isRequired={isRequired} title={title} />);

        expect(screen.getByTestId('checkbox-label')).toHaveClass('checkbox');
    });

    // is Required? aria-required
    test('Checkbox is required', () => {
        render(<Checkbox content={content} id={id} isRequired={isRequired} title={title} />);

        expect(screen.getByTestId('checkbox-input')).toBeRequired();
    });

    // Checkbox is not required
    test('Checkbox is not required', () => {
        render(<Checkbox content={content} id={id} title={title} />);

        expect(screen.getByTestId('checkbox-input')).not.toBeRequired();
    });

    // Checkbox is checked on click
    test('Checkbox is checked', () => {
        render(<Checkbox content={content} id={id} isRequired={isRequired} title={title} />);

        userEvent.click(screen.getByText(content));
        expect(screen.getByTestId('checkbox-input')).toBeChecked();
        expect(screen.getByTestId('checkbox-input-control')).toHaveAttribute('aria-checked');
    });
});
