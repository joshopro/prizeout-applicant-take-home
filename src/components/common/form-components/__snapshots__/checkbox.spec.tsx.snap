// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Checkbox Component Component matches snapshot 1`] = `
<DocumentFragment>
  <label
    class="checkbox"
    data-testid="checkbox-label"
    for="checkbox-test-id"
  >
    <input
      aria-required="true"
      class="checkbox__input"
      data-testid="checkbox-input"
      id="checkbox-test-id"
      title="Checkbox content"
      type="checkbox"
    />
    <span
      aria-checked="false"
      aria-labelledby="checkbox-test-id"
      class="checkbox__control"
      data-testid="checkbox-input-control"
    >
      <small
        data-testid="checkbox-description"
      >
        Checkbox content
      </small>
    </span>
  </label>
</DocumentFragment>
`;
